src/util/threadpool.h

```diff
@@ -0,0 +1,124 @@
+// Copyright (c) 2023-present The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or https://www.opensource.org/licenses/mit-license.php.
+
+#ifndef BITCOIN_UTIL_THREADPOOL_H
+#define BITCOIN_UTIL_THREADPOOL_H
+
+#include <sync.h>
+#include <util/string.h>
+#include <util/threadnames.h>
+
+#include <condition_variable>
+#include <future>
+#include <queue>
+#include <thread>
+
+class ThreadPool {
+
+private:
+    Mutex cs_work_queue;
+    std::queue<std::function<void()>> m_work_queue GUARDED_BY(cs_work_queue);
+    std::condition_variable m_condition;
+    // Stop indicator
+    std::atomic<bool> m_stop{false};
+    // Worker threads
+    std::vector<std::thread> m_workers;
+
+    void WorkerThread(const std::string& thread_name) EXCLUSIVE_LOCKS_REQUIRED(!cs_work_queue)
+    {
+        util::ThreadRename(std::string{thread_name});
+        WAIT_LOCK(cs_work_queue, wait_lock);
+        while (!m_stop.load()) {
+            std::function<void()> task;
+            {
+                // Wait for the task or until the stop flag is set
+                m_condition.wait(wait_lock,[&]() EXCLUSIVE_LOCKS_REQUIRED(cs_work_queue) { return m_stop.load() || !m_work_queue.empty(); });
+
+                // If stopped, exit worker.
+                if (m_stop.load() && m_work_queue.empty()) {
+                    return;
+                }
+
+                // Pop the task
+                task = std::move(m_work_queue.front());
+                m_work_queue.pop();
+            }
+
+            // Execute the task without the lock
+            WITH_REVERSE_LOCK(wait_lock, task());
+        }
```

Comment:
I don't think it would be necessary for the tasks themselves, since you already demonstrate in the tests how to retrieve exceptions, but I think some kind of exception handling and infomation logging similar to the one of `util::TraceThread` would still be nice. Did you choose not to use `TraceThread` on purpose?